api/sandbox.go

```diff
@@ -0,0 +1,114 @@
+package api
+
+import (
+	"context"
+	"syscall"
+
+	"github.com/hyperhq/runv/api"
+	ocispecs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+type ContainerConfig struct {
+	Id string
+
+	UGI    *UserGroupInfo
+	Mounts map[string]Mount
+
+	OciSpec ocispecs.Spec
+}
+
+type Mount struct {
+	StorageID string
+	Path      string
+	ReadOnly  bool
+}
+
```

Comment:
virtcontainer's `Mount`:

```Go
// Mount describes a container mount.
type Mount struct {
	Source      string
	Destination string

	// Type specifies the type of filesystem to mount.
	Type string

	// Options list all the mount options of the filesystem.
	Options []string

	// HostPath used to store host side bind mount path
	HostPath string

	// ReadOnly specifies if the mount should be read only or not
	ReadOnly bool
}
```

Reply:
@bergwolf could you elaborate on how `Mount` would reference a `Storage`? Who decides that a container's mount is referring to a specific storage definition?

Reply:
It is referenced by letting `OCI.Mount.Source` == `pb.Storage.Mountpoint` (or a subdirectory). The kata runtime API caller needs to craft proper `OCI.Mount.Source` to establish the mapping. It is a bit awkward than the original runv API but the change is needed because we want to use libcontainer in kata agent.

Reply:
This does not need to be specified at the API level, this is very OCI specific. The kata agent implementation in virtcontainers `kata_agent.go` can take care of translating whatever is expected by the Kata agent itself.

Reply:
@sboeuf we want to pass the OCI spec around through the API to better help kata runtime communicate with libcontainer. The OCI spec field in the agent's API is there for the same purpose -- so that we do not convert from OCI spec to kata spec and then back the OCI spec. Although we do need to make necessary modifications, the most of the OCI spec gets through untouched.

OTOH, the main purpose to have `sandbox.AddStorage()` is to support explicit hotplug and let upper layer prepare container resources (nics/volumes/rootfs etc.) in parallel. If we eventually decide not to pass the OCI spec around, we still need an `Mount` field in kata runtime's container spec to properly reference `Storage`.

Reply:
The difference is that the container rootfs is prepared outside of the runtime library. By letting upper layer initiate the preparation in parallel, we can save hundreds of milliseconds when starting a container (actual time of saving depends on which storage is used though).

If we hide them in sandbox/container creation flow as in current virtcontainers API, we can do nothing but waiting when preparing the container rootfs.

I can live with OCI spec annotation that is really not different from passing through API. IMHO it gives better type checking to pass it through API. Just my two cents though.